contracts/azorius/Azorius.sol

```diff
@@ -0,0 +1,463 @@
+
+/// @title Azorius - A Zodiac module that enables a voting agnostic proposal mechanism
+contract Azorius is Module, IAzorius {
+    bytes32 public constant DOMAIN_SEPARATOR_TYPEHASH =
+        0x47e79534a245952e8b16893a336b85a3d9ea9fa8c573f3d803afb92a79469218;
+    bytes32 public constant TRANSACTION_TYPEHASH =
+        0x72e9670a7ee00f5fbf1049b8c38e3f22fab7e9b85029e85cf9412f17fdd5c2ad;
+    uint256 public totalProposalCount; // Total number of submitted proposals
+    uint256 public timelockPeriod; // Delay between a proposal is passed and can be executed
+    address internal constant SENTINEL_STRATEGY = address(0x1);
+    mapping(uint256 => Proposal) internal proposals; // Proposals by proposal ID
+    mapping(address => address) internal strategies;
+
+    event ProposalCreated(
+        address strategy,
+        uint256 proposalId,
+        address proposer,
+        Transaction[] transactions,
+        string metadata
+    );
+    event TransactionExecuted(uint256 proposalId, bytes32 txHash);
+    event TransactionExecutedBatch(uint256 startIndex, uint256 endIndex);
+    event ProposalTimelocked(uint256 proposalId, uint256 timelockDeadline);
+    event AzoriusSetup(
+        address indexed initiator,
+        address indexed owner,
+        address indexed avatar,
+        address target
+    );
+    event EnabledStrategy(address strategy);
+    event DisabledStrategy(address strategy);
+    event TimelockPeriodUpdated(uint256 newTimelockPeriod);
+
+    constructor(
+        address _owner,
+        address _avatar,
+        address _target,
+        address[] memory _strategies,
+        uint256 _timelockPeriod
+    ) {
+        bytes memory initParams = abi.encode(
+            _owner,
+            _avatar,
+            _target,
+            _strategies,
+            _timelockPeriod
+        );
+        setUp(initParams);
+    }
+
+    function setUp(bytes memory initParams) public override initializer {
+        (
+            address _owner,
+            address _avatar,
+            address _target,
+            address[] memory _strategies,
+            uint256 _timelockPeriod
+        ) = abi.decode(initParams, (address, address, address, address[], uint256));
+        __Ownable_init();
+        avatar = _avatar;
+        target = _target;
+        setupStrategies(_strategies);
+        transferOwnership(_owner);
+        _updateTimelockPeriod(_timelockPeriod);
+
+        emit AzoriusSetup(msg.sender, _owner, _avatar, _target);
+    }
+
+    /// @notice Enables a voting strategy that can vote on proposals, only callable by the owner
+    /// @param _strategy Address of the strategy to be enabled
+    function enableStrategy(address _strategy) public onlyOwner {
+        require(
+            _strategy != address(0) && _strategy != SENTINEL_STRATEGY,
+            "Invalid strategy"
+        );
+        require(
+            strategies[_strategy] == address(0),
+            "Strategy already enabled"
+        );
```

Comment:
@sethhrbek does not using custom errors save on gas?  If so yeah I'd vote for just cutting them all

Reply:
@herbig --

Using custom errors with revert could potentially save on gas if we didn't use a revert string.

```
error InvalidStrategy();

if (_strategy == address(0) || _strategy == SENTINEL_STRATEGY) {
  revert InvalidStrategy();
}
```

If we could also pass an argument to the custom error to get more info (`error InvalidStrategy(address strategy)`).

The front-end would need to decode the custom error selector (see "errors with parameters" here) https://blog.soliditylang.org/2021/04/21/custom-errors/. Ethers has some updates since this article was written which should make catching the errors easier. 

There would be gas savings on deploy, but runtime gas savings would only happen if a revert is hit. Gas savings may be negligible enough to not warrant changing these, but thought it was worth a discussion.

Reply:
Yeah, I'm definitely interested in implementing these custom errors for Fractal. Maybe we can include these in this next PR if we think its worth it.